File: April2022/Java/EncodeAndDecodeTinyUrl.java

// @saorav21994
// TC : O(log <base 62> (number of test cases))
// SC : O(n)

// This question is more focussed on system design.

public class Codec {
    
    String baseString = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    char [] base = baseString.toCharArray();
    Map<String, Integer> shortMap = new HashMap<>();
    Map<Integer, String> longMap = new HashMap<>();
    static int id = 0;
    final int MAX = 62;
    
    // Encodes a URL to a shortened URL.
    public String encode(String longUrl) {
        
        id += 1;
        int copy = id;
        StringBuffer buffer = new StringBuffer();
        while (copy > 0) {
            buffer.append(base[copy%MAX]);
            copy /= MAX;
        }
        String shortUrl = buffer.reverse().toString();
        shortMap.put(shortUrl, id);
        longMap.put(id, longUrl);
        return shortUrl;
        
    }

    // Decodes a shortened URL to its original URL.
    public String decode(String shortUrl) {
     
        int id = shortMap.getOrDefault(shortUrl, 0);
        if (id == 0)
            return "";
        
        return longMap.getOrDefault(id, "");
        
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.decode(codec.encode(url));

File: April2022/Kotlin/EncodeandDecodeTinyUrl.kt
// Author: Shobhit Behl (LC: shobhitbruh)

class Codec {
    // Encodes a URL to a shortened URL.
    /*
    
    Here i have decoded the string by taking the int value of its each character 
    and by adding '.' in front of it.

    For e.g in string = "add" the encoded string would be "97.100.100."
    
    
    */
    fun encode(longUrl: String): String {
        val l = longUrl.toCharArray()
        val sb = StringBuilder()
        for (i in l.indices) {
            val k = l[i].toInt()
            sb.append(k)
            sb.append('.')
        }
        return sb.toString()
    }

    // Decodes a shortened URL to its original URL.
    fun decode(shortUrl: String): String {
        val y = shortUrl.split("\\.".toRegex()).toTypedArray()
        val sb = StringBuilder()
        for (i in y.indices) {
            if(y[i].length==0){
                break;
            }
            val k:Int = y[i].toInt()
            val p = k.toChar()
            sb.append(p)
        }
        return sb.toString()
    }
}